### DISCUSSION TITLE
High VRAM usage with many lights / Expose max light count to UI

### DISCUSSION THREAD

**Discussion by VelocityRa** [bnpr/Malt#438] [Q&A]
A large scene I have with only a Sun light enabled as a light source uses about 1.8GB of VRAM and renders in ~0.8s.
As soon as I enable a collection of just 20 point lights, VRAM usage jumps to ~8GB and render time to 2s (not interactive anymore). Not sure I can feasibly add any more lights without things breaking.

My rendering resolution is only 192x192, I don't need much detail in the lighting/shadows, is there a way I can make the lights use less resources? Not sure if a shadow mapping based approach is used here, maybe I can reduce the shadow map resolution or something.

**Comment by VelocityRa** [comment]
> maybe I can reduce the shadow map resolution or something

Seems so https://github.com/bnpr/Malt/blob/Development/Malt/Render/Lighting.py#L67-L80.

Not sure if this is the cause of the high VRAM usage, but it might be useful to expose these to the UI.
Ditto for the max allowed light counts (you can presumably have lots more with the same amount of VRAM if res is reduced).
Changed the post title.

_Edit_: Might not the cause, see below.

**Reply by VelocityRa** [reply]
Very back of the envelope calculation but: current res of the spotlights is 2K, `2048*2048*20*4` (20 is the count, 4 is the number of bytes in a float) = `320MB`. This does not nearly account for the ~`6GB` VRAM usage increase so not sure what's up.
I'm guessing there's more textures/data involved in calculating the lighting that are of significant size.
Perhaps something per-mesh? Just a guess, since shadows are handled in Mesh shaders and I have hundreds.

**Reply by VelocityRa** [reply]
Hm. I lowered `point_resolution` to 64 from 512 and it made no difference to VRAM usage.
I later confirmed my changes were being applied by injecting an error in `Lighting.py` (Malt didn't work as expected)
I restarted Blender after every change.

**Reply by pragma37** [reply]
Here's where you should change the shadow map settings:
![imagen](https://user-images.githubusercontent.com/36610999/199035187-481d0dc9-df3f-4250-8a7c-c54a6156349f.png)
(You may need to reload the pipeline for the vram to actually shrink)
There's currently no way to simply disable shadows for certain lights, but it should be added.

> I'm guessing there's more textures/data involved in calculating the lighting that are of significant size

For each light, 2 depth maps (1 for opaque shadows, 1 for transparent), 2 ID maps (for detecting self-shadows in opaque and transparent shadows) and 1 RGB8 map for transparent shadows colors.

**Reply by VelocityRa** [reply]
Very interesting, thanks!
Didn't know these were exposed already and saved in the blend.
I reduced these to half/quarter and my scene now only uses ~600MB more VRAM instead of 7GB with the 20 spotlights :)
